### PR TITLE
Added object path expressions #808 #693

### LINKED ISSUE #693: Add support for field names that enumerate an object path
Current field names can index into an object or array, for example `Property[0].name`. Only a single field can be returned.

Allow more then one field to be return by enumerating an object. i.e. `Property[*].name`.

### LINKED ISSUE #808: Add support for JsonPath queries syntax in fields
Instead of using simple fields such as `Property[0].name` allow fields to be queried using JsonPath syntax.

Related to #693.

### PR DESCRIPTION
**Pull request by BernieWhite**
## PR Summary

- Added support for object path expressions.
    - Inspired by JSONPath, object path expressions can be used to access nested objects.
    - Array members can be filtered and enumerated using object path expressions.
    - Object path expressions can be used in YAML, JSON, and PowerShell rules and selectors.

Fixes #808 
Fixes #693

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Code changes**
  - [x] Have unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Microsoft/PSRule/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule/Runtime/ObjectPath/PathExpressionBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+
+            value = fieldInfo.GetValue(targetObject);
+            return true;
+        }
+
+        private static bool TryIndexerProperty(object targetObject, object index, Type baseType, out object value)
+        {
+            value = null;
+            var properties = baseType.GetProperties();
+            foreach (PropertyInfo pi in GetIndexerProperties(baseType))
```
[the remark this answers is not in the document]
Comment (ArmaanMcleod) [reply src/PSRule/Runtime/ObjectPath/PathExpressionBuilder.cs:588]: Maybe `var property` would be more clearer?

File: src/PSRule/Runtime/ObjectPath/PathExpressionBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+            var attribute = baseType.GetCustomAttribute<DefaultMemberAttribute>();
+            if (attribute != null)
+            {
+                var property = baseType.GetProperty(attribute.MemberName);
+                yield return property;
+            }
+            else
+            {
+                var properties = baseType.GetProperties();
+                foreach (PropertyInfo pi in properties)
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/PathExpressionBuilder.cs:640]:
```suggestion
                foreach (var pi in properties)
```

File: src/PSRule/Runtime/ObjectPath/PathTokenizer.cs [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+            }
+
+            public IPathToken Last
+            {
+                get
+                {
+                    if (_Items.Count == 0)
+                        return null;
+
+                    return _Items[_Items.Count - 1];
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/PathTokenizer.cs:31]:
```suggestion
                    return _Items.Count > 0 ? _Items[_Items.Count - 1] : null;
```

File: src/PSRule/Runtime/ObjectPath/PathTokenizer.cs [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+                    tokens.Add(token);
+                }
+            }
+
+            private char ConsumeChar(ref int position)
+            {
+                if (position > Last)
+                    return NULL;
+
+                return Path[position++];
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/PathTokenizer.cs:296]:
```suggestion
                return position > Last : NULL ? Path[position++];
```

File: src/PSRule/Runtime/ObjectPath/PathTokenizer.cs [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+                position = pos;
+                return true;
+            }
+
+            private string CaptureMemberName(ref int position)
+            {
+                if (!UntilQuote(ref position, out string value) && !WhileMemberName(ref position, out value))
+                    return null;
+
+                return value;
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/PathTokenizer.cs:541]:
```suggestion
                return UntilQuote(ref position, out string value) && !WhileMemberName(ref position, out value) ? value : null;
```

File: src/PSRule/Runtime/ObjectPath/PathTokenizer.cs [hunk 6]
```diff
@@ -0,0 +1,695 @@
...
+
+                return pos;
+            }
+
+            private string Substring(int pos, int end)
+            {
+                if (pos == end)
+                    return null;
+
+                return Path.Substring(pos, end - pos);
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/PathTokenizer.cs:597]:
```suggestion
                return pos == end ? null : Path.Substring(pos, end - pos);
```

File: src/PSRule/Runtime/ObjectPath/Tokens.cs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        public IPathToken Current { get; private set; }
+
+        public bool Consume(PathTokenType type)
+        {
+            if (Peak(out IPathToken token) && token.Type == type)
+                return Next();
+
+            return false;
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/ObjectPath/Tokens.cs:173]:
```suggestion
            return (Peak(out IPathToken token) && token.Type == type) ? Next() : false;
```

File: src/PSRule/Runtime/PSRule.cs [hunk 8]
```diff
@@ -189,6 +189,17 @@ public PSObject GetContentFirstOrDefault(PSObject sourceObject)
...
 
+        /// <summary>
+        /// Evalute an object path expression and returns the resulting objects.
+        /// </summary>
+        public object[] GetPath(object sourceObject, string path)
+        {
+            if (!ObjectHelper.GetPath(GetContext()?.Pipeline, sourceObject, path, false, out object[] value))
+                return Array.Empty<object>();
+
+            return value;
```
Comment (ArmaanMcleod) [inline src/PSRule/Runtime/PSRule.cs:200]:
```suggestion
            return (!ObjectHelper.GetPath(
                bindingContext: GetContext()?.Pipeline, 
                targetObject: sourceObject, 
                path: path, 
                caseSensitive: false, 
                out object[] value)) ? Array.Empty<object>() : value;
```

**Review by ArmaanMcleod — APPROVED** [review]
This looks really good @BernieWhite 👏 , nice work. I added some small code cleanup suggestions, not sure if it makes things more readable but it seems easier for me to read. No need to add them in if you don't see benefit or much gain in readability 🙂. 

Also happy new year mate 🎉 🎉 🎉

**Comment by BernieWhite** [comment]
@ArmaanMcleod Thanks. Happy new year.

**Pull request merged**